### PR TITLE
Stdlib.Format: add pp_print_seq

### PR DESCRIPTION
**Pull request by raphael-proust**
As the title suggest, this PR adds `pp_print_seq` to the `Format` module of the `Stdlib`. I stumbled upon a situation where I needed to convert a perfectly fine `Seq.t` into a `List.t` for the sole purpose of printing. So I made this PR.

I have not tested it yet, but I'd be happy to do so (as well as to pour some more energy/time) if it's considered to be a relevant PR that may lead to being merged.

### CODE REVIEW

**Comment by ghost** [comment]
Could you tell us more about that conversion situation?

**Comment by shindere** [comment]
In general the stdlib is not my area of expertise, so I'll wait for
other more involved developers to comment, but a priori I'd say it would
be nice to have this function added to the format module.

**Comment by xvw** [comment]
To be honnest, I think this function is pretty useful. I often want to print sequences and I always go through the conversion to lists. Even for reasons other than debugging (using printf). And in general, I find it pretty good to have pretty-printers for the usual data structures of the standard library!

**Comment by shindere** [comment]
Xavier Van de Woestyne (2020/08/31 06:02 -0700):
> To be honnest, I think this function is pretty useful. I often want to
> print sequences and I always go through the conversion to lists. Even
> for reasons other than debugging (using printf). And in general, I
> find it pretty good to have pretty-printers for the usual data
> structures of the standard library!

I fully agree.

**Comment by raphael-proust** [comment]
> Could you tell us more about that conversion situation?

I have a hashtbl and I need to print the keys. I'd rather use a `pp_print_seq` on a `Hashtbl.to_seq_keys` (i.e., iterating once through a lazy iterator on the underlying structure) than either (1) iterating over the hashtbl by hand (I don't want to handle the printing of the separator), (2) converting the seq to a list and printing that (why allocate the list?), (3) building the list directly by folding over the table (same reasons).

File: stdlib/format.mli
Comment (nojb) [inline stdlib/format.mli]:
```suggestion
  @since 4.12
```

**Review by nojb — COMMENTED** [review]
LGTM. Thanks!

This PR needs a test and a Changes entry.

Also, let's wait a few more days to give other developers time to chime in.

**Comment by nojb** [comment]
Also, maybe indicate in the docstring that the function does not terminate if the sequence does not terminate.

**Comment by shindere** [comment]
Will you please post a comment here to let us know when you feel the code is ready for review?

**Comment by raphael-proust** [comment]
I have now rebased (squashed all the commits together). This PR is ready for review.

**Comment by shindere** [comment]
The code looks nice to me, thanks!

You may want to credit your reviewers in Changes (the real ones I mean,
not me).

Would you be willing to include some tests?

**Comment by raphael-proust** [comment]
I'm ok adding credit to reviewers, I'm just unfamiliar with the process here. I guess @nojb 's comments constitute a review. Will there be other reviews (in which case, should I wait for these reviews and add all in one go?) or is this basically reviewed?

**Comment by Octachron** [comment]
Crediting just the original reviewer (aka @nojb ) is fine.

(Also as another developer data point, I also think that `pp_print_seq` is a legitimate addition to Format.)

File: testsuite/tests/lib-format/print_seq.ml
Comment (nojb) [inline testsuite/tests/lib-format/print_seq.ml:18]: This is not valid syntax. You can preface the expression with `let () = try ...` instead.

**Comment by shindere** [comment]
My understan ding is that you add to the Changes entry the people you
feel grateful to because they somehow helped. I don't know whether there
will be other reviews or not. The PR needs at least one approval to be
mergeable.

File: testsuite/tests/lib-format/print_seq.ml
Comment (raphael-proust) [reply testsuite/tests/lib-format/print_seq.ml:18]: I use the style from another test in the same directory, and it uses `;;` termination rather than `let () =` at top level. I made a mistake when adding this intermediate function.

It's fixed now. I also added a corresponding `reference` file and I ran the test locally.

**Review by nojb — APPROVED** [review]
LGTM

Good to merge once CI passes.

File: testsuite/tests/lib-format/print_seq.ml
Comment (Octachron) [inline testsuite/tests/lib-format/print_seq.ml]: This blank line should be be removed in order to pass the `check-typo` test.

Comment (raphael-proust) [reply testsuite/tests/lib-format/print_seq.ml]: Thanks! I fixed it and ran the test locally to check.

**Pull request merged**
